fili-core/src/main/java/com/yahoo/bard/webservice/web/filterparser/ApiFiltersListListener.java

```diff
@@ -0,0 +1,222 @@
+     * @return the DimensionField
+     */
+    protected DimensionField extractField(FiltersParser.FilterComponentContext f, Dimension dimension) {
+        String dimensionFieldName;
+        dimensionFieldName = f.field().getText();
+        DimensionField dimensionField;
+        try {
+            dimensionField = dimension.getFieldByName(dimensionFieldName);
+        } catch (IllegalArgumentException ignored) {
+            LOG.debug(FILTER_FIELD_NOT_IN_DIMENSIONS.logFormat(dimensionFieldName, dimension.getApiName()));
+            errors.add(new BadFilterException(
+                    FILTER_FIELD_NOT_IN_DIMENSIONS.format(dimensionFieldName, dimension.getApiName()))
+            );
+            return null;
+        }
+        return dimensionField;
+    }
+
+    /**
+     * Extract the dimension portion of a filter.
+     * @param f The filter context
+     * @return the Dimension
+     */
+    protected Dimension extractDimension(FiltersParser.FilterComponentContext f) {
+        String dimensionName = f.dimension().getText();
+        Dimension dimension = dimensionDictionary.findByApiName(dimensionName);
+
+        // If no filter dimension is found in dimension dictionary throw exception.
+        if (dimension == null) {
+            LOG.debug(FILTER_DIMENSION_UNDEFINED.logFormat(dimensionName));
+            errors.add(new BadFilterException(FILTER_DIMENSION_UNDEFINED.format(dimensionName)));
+            return null;
+        }
+
+        // If there is a logical table and the filter is not part of it, throw exception.
+        if (table != null && !table.getDimensions().contains(dimension)) {
+            LOG.debug(FILTER_DIMENSION_NOT_IN_TABLE.logFormat(dimensionName, table));
+            errors.add(new BadFilterException(
+                    FILTER_DIMENSION_NOT_IN_TABLE.format(dimensionName, table.getName()))
+            );
+            return null;
+        }
+        return dimension;
+    }
+
+    /**
+     * Gets the list of parsed dimensionFiltersMap. If a parsing error occured or an invalid filter was
+     * specified, throws BadFilterException
+     *
+     * @return a Map of ApiFilter keyed by Dimension
+     * @throws BadFilterException Thrown when an invalid filter is specified
+     * @throws BadApiRequestException Thrown when a filter isn't supported
+     */
+    public Map<Dimension, Set<ApiFilter>> getDimensionFiltersMap () throws BadFilterException, BadApiRequestException {
+        processErrors();
+        return dimensionFiltersMap;
+    }
+
+    /**
+     * Throw any pending exceptions.
+     *
+     * @throws BadFilterException Thrown when the filter is invalid
+     * @throws BadApiRequestException Thrown for valid filters that are invalid for the current table or dimensions
+     */
+    protected void processErrors() throws BadFilterException, BadApiRequestException {
+        for (Exception ex : errors) {
+            if (ex instanceof BadFilterException) {
```

Comment:
If we're only returning the first error, why not bail out of parsing?  Should we be collecting all the errors into a new error and then returning it for maximum user feedback?